utils/net_utils/src/download.rs

```diff
@@ -0,0 +1,103 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use async_std::{
+    fs,
+    io::{self, copy, Read as AsyncRead},
+};
+use futures::task::{Context, Poll};
+use indicatif::{ProgressBar, ProgressStyle};
+use isahc::HttpClient;
+use log::info;
+use std::{marker::Unpin, path::Path, pin::Pin};
+use thiserror::Error;
+use url::Url;
+
+/// Contains progress bar and reader.
+struct DownloadProgress<S>
+where
+    S: Unpin,
+{
+    inner: S,
+    progress_bar: ProgressBar,
+}
+
+#[derive(Debug, Error)]
+enum DownloadError {
+    #[error("Cannot read a file header")]
+    HeaderError,
+    #[error("Filename encoding error")]
+    EncodingError,
+}
+
+impl<S: AsyncRead + Unpin> AsyncRead for DownloadProgress<S> {
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut [u8],
+    ) -> Poll<Result<usize, io::Error>> {
+        match Pin::new(&mut self.inner).poll_read(cx, buf) {
+            Poll::Ready(Ok(size)) => {
+                if size == 0 {
+                    self.progress_bar.finish();
```

Comment:
This case won't be hit (unless I'm missing something here)

Reply:
not relevant in a new iteration of code.